rust/otel-arrow-rust/src/encode/record/array/dictionary.rs

```diff
@@ -350,25 +350,24 @@ where
     T16: DictionaryArrayAppend<Native = T> + ArrayBuilderConstructor,
 {
     pub fn append_value(&mut self, value: &T) -> Result<usize> {
-        let append_result = match &mut self.variant {
-            DictIndexVariant::UInt8(dict_builder) => dict_builder.append_value(value),
-            DictIndexVariant::UInt16(dict_builder) => dict_builder.append_value(value),
-        };
-
-        match append_result {
-            Ok(index) => {
-                if index + 1 > self.max_cardinality as usize {
-                    // if we're here, it means we did append successfully to the underlying builder
-                    // but we shouldn't have, because have overflowed the configured max cardinality
-                    self.overflow_index = Some(index);
-                    Err(DictionaryBuilderError::DictOverflow {})
-                } else {
-                    Ok(index)
+        loop {
+            let append_result = match &mut self.variant {
+                DictIndexVariant::UInt8(dict_builder) => dict_builder.append_value(value),
+                DictIndexVariant::UInt16(dict_builder) => dict_builder.append_value(value),
+            };
+
+            match append_result {
+                Ok(index) => {
+                    if index + 1 > self.max_cardinality as usize {
+                        self.overflow_index = Some(index);
+                        return Err(DictionaryBuilderError::DictOverflow {});
+                    }
+                    return Ok(index);
+                }
+                Err(DictionaryBuilderError::DictOverflow {}) => {
+                    self.upgrade_key()?;
+                    // retry in the next loop iteration
                 }
-            }
-            Err(DictionaryBuilderError::DictOverflow {}) => {
-                self.upgrade_key()?;
-                self.append_value(value)
```

Comment:
turns out using a loop instead of recursive call to append here is a good performance improvement